### PR TITLE
allow a verifyoptionsfunc to indicate that no certpool is available

### PR DESCRIPTION
**Pull request by deads2k**
xref https://github.com/kubernetes/kubernetes/issues/82141

Adds bool to VerifyFunc.  If the bool is false, then the returned VerifyOptions are ignored and the authenticator will express "no opinion".  This allows a clear signal for cases where a CertPool is eventually expected, but not currently present.

/kind bug
/priority important-soon
@kubernetes/sig-auth-pr-reviews 

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_cafile_content.go [hunk 1]
```diff
@@ -215,8 +215,17 @@ func (c *DynamicFileCAContent) CurrentCABundleContent() (cabundle []byte) {
...
 
 // VerifyOptions provides verifyoptions compatible with authenticators
-func (c *DynamicFileCAContent) VerifyOptions() x509.VerifyOptions {
-	return c.caBundle.Load().(*caBundleAndVerifier).verifyOptions
+func (c *DynamicFileCAContent) VerifyOptions() (x509.VerifyOptions, bool) {
+	uncastObj := c.caBundle.Load()
+	if uncastObj == nil {
+		return x509.VerifyOptions{}, false
+	}
+	castObj, ok := uncastObj.(*caBundleAndVerifier)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_cafile_content.go]: the type of an atomic.Value can't be changed once set. if we have a non-nil value that isn't the correct type, that's a programming error we should hard-fail on

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/union_content.go [hunk 2]
```diff
@@ -55,7 +55,12 @@ func (c unionCAContent) CurrentCABundleContent() []byte {
 }
 
 // CurrentCABundleContent provides ca bundle byte content
-func (c unionCAContent) VerifyOptions() x509.VerifyOptions {
+func (c unionCAContent) VerifyOptions() (x509.VerifyOptions, bool) {
+	currCABundle := c.CurrentCABundleContent()
+	if len(currCABundle) == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/union_content.go]: does this mean it is not possible to express "use system roots"?

File: test/integration/scheduler/taint_test.go [hunk 3]
```diff
@@ -656,6 +656,8 @@ func TestTaintBasedEvictions(t *testing.T) {
 	for i, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
 			context := initTestMaster(t, "taint-based-evictions", admission)
+			defer context.closeFn()
```
Comment (liggitt) [inline test/integration/scheduler/taint_test.go]: unrelated?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/union_content.go (hunk 2 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/union_content.go]: > does this mean it is not possible to express "use system roots"?

I'm pretty sure that we explicitly chose to have that *not* be the meaning of empty for client certificate validation, since that would imply that any officially signed client cert in the world is a valid identity for a given cluster instead of selectively choosing.  On the server verification side, it's very important.  On the client verification side, I question its importance and whether we ever supported it.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/union_content.go]: I think you're right

File: test/integration/scheduler/taint_test.go (hunk 3 above)
Comment (deads2k) [reply test/integration/scheduler/taint_test.go]: > unrelated?

found it working on this one, just a forgotten close

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_cafile_content.go (hunk 1 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_cafile_content.go]: > the type of an atomic.Value can't be changed once set. if we have a non-nil value that isn't the correct type, that's a programming error we should hard-fail on

updated to panic

**Comment by deads2k** [comment]
comments addressed
